### PR TITLE
[Caffe2] Support non peer access in muji and fix bug when reduced_affix is empty

### PR DESCRIPTION
**Pull request by daquexian**
reference: https://github.com/facebookresearch/Detectron/issues/32

Detectron users with gpus lacking peer access can't train models using multiple gpus. A workaround is using nccl but it may [suffer from deadlock](https://github.com/facebookresearch/Detectron/issues/32#issuecomment-361140510).

I updated the code in muji so that it will use the proper way of allreduce according to the peer access pattern. Since there are many users having four gpus where pear access are enabled between {0,1} and {2,3}(e.g. myself), I also added an `Allreduce4Group2` function for this case.

I have verified it on [configs/getting_started/tutorial_4gpu_e2e_faster_rcnn_R-50-FPN.yaml](https://github.com/facebookresearch/Detectron/blob/master/configs/getting_started/tutorial_4gpu_e2e_faster_rcnn_R-50-FPN.yaml) of detectron

### CODE REVIEW

**Comment by daquexian** [comment]
I don't think there is such a bug. Do you come across a bug in this case?

**Comment by daquexian** [comment]
@yuzcccc I have reproduced it on my side. It seems a inherent bug in `AllreduceFallback`

**Comment by daquexian** [comment]
@yuzcccc I have fixed it. Please try the updated code.

**Comment by daquexian** [comment]
@orionr Could you please review it? Thanks! :)

**Comment by daquexian** [comment]
@Yangqing Could anyone review it? :)

**Comment by pjh5** [comment]
@daquexian judging from the error messages in those conda builds, you are based off code that's a few weeks old. Can you rebase off of master and push again?

**Comment by daquexian** [comment]
@pjh5 Thanks for your advice! Now all checks passed! 

@orionr @ezyang This PR ready to review now :)

File: caffe2/python/muji.py
Comment (orionr) [inline caffe2/python/muji.py]: Nit: satisified -> satisfied

File: caffe2/python/muji.py
Comment (orionr) [inline caffe2/python/muji.py]: Any chance this method could be called on CPU? Seems like the answer is no, but confirming. Also, we will eventually be supporting HIP/AMD, so I wonder if this is safe. cc @bddppq for that.

**Review by orionr — COMMENTED** [review]
Looks good! My only concern is around CPU execution and HIP/AMD. I'll wait for @bddppq's quick review before accepting.

File: caffe2/python/muji.py
Comment (bddppq) [reply caffe2/python/muji.py]: The plan for AMD multi-gpu support is still unsettled, and very likely we will need to special case/fork the code anyway, so I would say let's don't consider that for now.

**Review by orionr — APPROVED** [review]
Only the typo nit and question around whether this can be called on CPU. Approving to unblock, but please comment on those before landing. Thanks.

**Comment by daquexian** [comment]
@orionr Thanks for your approving! I have fixed the typo and added comments.

**Comment by daquexian** [comment]
@orionr Is it ready to land it? Thanks!

**Pull request merged**
